dandi/misctypes.py

```diff
@@ -0,0 +1,241 @@
+"""
+.. versionadded:: 0.36.0
+
+Miscellaneous public classes
+"""
+
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from fnmatch import fnmatchcase
+from typing import Dict, Iterator, List, Tuple, TypeVar
+
+from dandischema.models import DigestType
+
+
+@dataclass
+class Digest:
+    """A computed digest for a file or directory"""
+
+    #: The digest algorithm used
+    algorithm: DigestType
+
+    #: The digest itself
+    value: str
+
+    @classmethod
+    def dandi_etag(cls, value: str) -> Digest:
+        """
+        Construct a `Digest` with the given value and a ``algorithm`` of
+        ``DigestType.dandi_etag``
+        """
+        return cls(algorithm=DigestType.dandi_etag, value=value)
+
+    @classmethod
+    def dandi_zarr(cls, value: str) -> Digest:
+        """
+        Construct a `Digest` with the given value and a ``algorithm`` of
+        ``DigestType.dandi_zarr_checksum``
+        """
+        return cls(algorithm=DigestType.dandi_zarr_checksum, value=value)
+
+    def asdict(self) -> Dict[DigestType, str]:
+        """
+        Convert the instance to a single-item `dict` mapping the digest
+        algorithm to the digest value
+        """
+        return {self.algorithm: self.value}
+
+
+#: Placeholder digest used in some situations where a digest is required but
+#: not actually relevant and would be too expensive to calculate
+DUMMY_DIGEST = Digest(algorithm=DigestType.dandi_etag, value=32 * "d" + "-1")
+
+P = TypeVar("P", bound="BasePath")
+
+
+@dataclass
+class BasePath(ABC):
+    """
+    An abstract base class for path-like objects that can be traversed with the
+    ``/`` operator *à la* `pathlib.Path` (though, unlike `pathlib.Path`
+    instances, "dividing" by another non-string path is not allowed).  All
```

Comment:
if we had to disable some "dividing" couldn't we just subclass `pathlib.Path` and sensor the divider?  otherwise feels like PosixPath could have been a good fit and avoided "reimplementing" it here?

Reply:
I'm not clear on what you're proposing in your first sentence, but the whole point of this class is to enable traversal with `/`.  The concrete subclasses require the instances to be aware of what resources they're pointing to, so replacing them with relative `PurePosixPath` instances wouldn't work, and subclassing `PurePosixPath` just seems like a recipe for implementation-defined disaster.

Reply:
ok, thank you!